python/lsst/pipe/base/pipeline_graph/_pipeline_graph.py

```diff
@@ -1524,6 +1539,337 @@ def split_independent(self) -> Iterable[PipelineGraph]:
+        `QuantumGraph`.  Use `QuantumGraph.write_init_outputs` if a quantum
+        graph has already been built.
+        """
+        init_outputs: list[tuple[Any, DatasetType]] = []
+        self.instantiate_tasks(butler.get, init_outputs)
+        found_refs: dict[str, DatasetRef] = {}
+        to_put: list[tuple[Any, DatasetType]] = []
+        for obj, dataset_type in init_outputs:
+            if (ref := butler.find_dataset(dataset_type, collections=butler.run)) is not None:
+                found_refs[dataset_type.name] = ref
+            else:
+                to_put.append((obj, dataset_type))
+        for ref, stored in butler.stored_many(found_refs.values()).items():
+            if not stored:
+                raise FileNotFoundError(
+                    f"Init-output dataset {ref.datasetType.name!r} was found in RUN {ref.run!r} "
+                    f"but had not actually been stored (or was stored and later deleted)."
+                )
+        for obj, dataset_type in to_put:
+            butler.put(obj, dataset_type)
+
+    def write_configs(self, butler: Butler) -> None:
+        """Write the config datasets for all tasks in the pipeline graph.
+
+        Parameters
+        ----------
+        butler : `lsst.daf.butler.Butler`
+            A full butler data repository client with its default run set
+            to the collection where datasets should be written.
+
+        Notes
+        -----
+        Config datasets that already exist in the butler's output run
+        collection will be checked for consistency.
+
+        This method writes outputs with new random dataset IDs and should
+        hence only be used when writing init-outputs prior to building a
+        `QuantumGraph`.  Use `QuantumGraph.write_configs` if a quantum graph
+        has already been built.
+
+        Raises
+        ------
+        lsst.daf.butler.registry.ConflictingDefinitionError
+            Raised if an config dataset already exists and is not consistent
+            with the config in the pipeline graph.
+        """
+        to_put: list[tuple[PipelineTaskConfig, str]] = []
+        for task_node in self.tasks.values():
+            dataset_type_name = task_node.init.config_output.dataset_type_name
+            if (ref := butler.find_dataset(dataset_type_name, collections=butler.run)) is not None:
+                old_config = butler.get(ref)
+                if not task_node.config.compare(old_config, shortcut=False, output=log_config_mismatch):
+                    raise ConflictingDefinitionError(
+                        f"Config does not match existing task config {dataset_type_name!r} in "
+                        "butler; tasks configurations must be consistent within the same run collection"
+                    )
+            else:
+                to_put.append((task_node.config, dataset_type_name))
+        # We do writes at the end to minimize the mess we leave behind when we
+        # raise an exception.
```

Comment:
I wondered about that too, but I think the extra complexity involved just isn't worth it, since this kind of mess hasn't really been a problem in practice that I'm aware of.